# Analysis: How does Discover change a customer's purchasing behaviour?

#### Context: 
Previously, visitors of PCSG website may feel overwhelmed/at loss by the sheer amount of products listed. Having a product recommendation system aims to reduce the friction to navigation/purchase, thereby increasing sales.
    
### Methodology
1. Import and format Discover and Shopify sales data
2. Combine the details of each Discover taker's **earliest** attempt into Shopify data
3. Identify NEW customers and match their earliest purchase with their earliest Discover attempt
4. Perform calculations, such as:
    1. Percentage of transactions that actually contain at least one of the recommended SKUs
    2. Average/median/min/max time taken to purchase after attempting Discover
    3. etc

### Note: 
1. This excludes: 
    1. All the customers who used Discover but DID NOT make a purchase.
    2. New Customers who purchased **without trying Discover** (either because they didn't bother, or didn't know of its existence)
    3. All customers who made purchases before Discover was launched
2. We do not have a "control group", i.e.: the time taken for pre-Discover new customers to make a purchase after browsing the website
3. We could potentially investigate the behaviour of those who made their first purchase WITHOUT trying out Discover (i.e. point 2), but subsequently tried Discover)


**note to self: since this is a "partial" dataset, I will need to rerun this experiment with the full dataset**

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
pd.set_option("display.max_columns", 100)

discover_data = pd.read_csv("discover_user_input_results.csv")
discover_data = discover_data[discover_data['email'] != 'jeremy@paulaschoice.sg']

# Obtain only the earliest attempt of each Discover taker
discover_first = discover_data.drop_duplicates(subset = ['email'], keep='first').copy()
discover_first['timestamp_proper'] = pd.to_datetime(discover_first['timestamp'].map(lambda x: x.replace('th', "")))
discover_first.drop('timestamp', axis=1,inplace=True)
discover_first.head()
# discover_first.to_csv('discover_first.csv')


Unnamed: 0                       email  sensitivity     skinType  \
2         937       ck1411@singnet.com.sg        False  Combination   
4         939       starlites18@gmail.com        False  Combination   
5         940  angelleexinyin@hotmail.com         True  Combination   
6         941     leong_lorna@hotmail.com         True         Oily   
7         942      xiaofangtm89@gmail.com        False  Combination   

        concern_0      concern_1       concern_2       concern_3  \
2  Enlarged Pores     Sun Damage        Wrinkles     Dehydration   
4  Enlarged Pores  Clogged Pores            Acne      Sun Damage   
5            Acne  Clogged Pores  Uneven Texture  Enlarged Pores   
6   Clogged Pores           Acne             PIH  Enlarged Pores   
7  Enlarged Pores  Clogged Pores            Acne  Uneven Texture   

          concern_4    concern_5    concern_6    concern_7         concern_8  \
2  Loss of Firmness     Dullness  Combination          NaN               NaN   
4    Uneven Texture      Redness  Combination          NaN               NaN   
5               PIH  Dehydration  Sensitivity  Combination               NaN   
6    Uneven Texture  Sensitivity         Oily          NaN               NaN   
7           Redness  Dehydration     Dullness          PIH  Loss of Firmness   

     concern_9 concern_10 concern_11 concern_12 concern_13  result_0  \
2          NaN        NaN        NaN        NaN        NaN      7830   
4          NaN        NaN        NaN        NaN        NaN      6002   
5          NaN        NaN        NaN        NaN        NaN      7830   
6          NaN        NaN        NaN        NaN        NaN      6002   
7  Combination        NaN        NaN        NaN        NaN      7830   

   result_1  result_2  result_3  result_4  result_5  result_6  result_7  \
2      7780      7820      7740      7870      7760      7690      2760   
4      1350      2010      7740      7870      6130      7690      2750   
5      1350      6200      7740      7870      6130      7690      2750   
6      1350      6200      7740      7870      6130      7690      2750   
7      7670      2010      7740      7870      6130      7690      2760   

   result_8  result_9    timestamp_proper  
2      7900      7960 2017-09-09 14:38:56  
4      5700      7730 2017-09-09 14:59:07  
5      5700      7730 2017-09-09 15:17:38  
6      5700      7730 2017-09-09 15:38:11  
7      5700      7730 2017-09-09 15:39:21

In [5]:
''' 
Sales Data
1. Import sales_data (dated 2018-02-07) and perform basic formatting 
2. Match the emails available from Discover dataset to sales data, and create a new column called "discover_first_date" to indicate when did the customer first use Discover
    Note: doing this excludes all the customers who used Discover but DID NOT make a purchase.
3. Create a second new column to indicate whether a transaction is made BEFORE/AFTER the customer has tried Discover.
    Note: customers that have not tried Discover *at all* but made a purchase will also show "Not Yet"
    Perhaps those who have not tried Discover at all should have a third status ("Not at all")?
4. Create a third column to indicate the time taken between the customer's first attempt of Discover and their current transaction.
'''

# Import Shopify sales_data (dated 2018-02-07) and perform basic formatting 
sales_data = pd.read_csv("shopify_orders_export_20180207.csv", 
                         low_memory=False, 
                         parse_dates=['Paid at', 'Fulfilled at', 'Created at'])

sales_data_clean = sales_data.drop(sales_data.columns.to_series()[-11:-1], axis=1)
sales_data_clean.dropna(subset=['Email'], axis=0, inplace=True)


# Match the emails available from Discover dataset to sales data, 
# and create a new column called `"discover_first_date"` to indicate 
# if and when customer first used Discover
sales_data_clean['discover_first_date'] = sales_data_clean['Email'].map(discover_first.set_index('email')['timestamp_proper'])

# Calculate `'used_discover_already'` : whether a transaction is made 
# BEFORE/AFTER the customer has tried Discover.
sales_data_clean['used_discover_already'] = (sales_data_clean['Created at']> sales_data_clean['discover_first_date']).map({True: "Used Discover", False: "Not yet"})

# Just to calculate the lead time to sales after trying Discover
sales_data_clean['discover_sales_lead_time'] = sales_data_clean['Created at'] - sales_data_clean['discover_first_date']

sales_data_clean.head(10)

Name                    Email Financial Status             Paid at  \
0  191569914712   gilly.glanville@me.com             paid 2018-02-08 03:59:00   
1  191569914712   gilly.glanville@me.com              NaN                 NaT   
2  191569914712   gilly.glanville@me.com              NaN                 NaT   
3  191569914712   gilly.glanville@me.com              NaN                 NaT   
4  191569914712   gilly.glanville@me.com              NaN                 NaT   
5  191569914711  stephaniedata@yahoo.com             paid 2018-02-08 03:39:25   
6  191569914710     hsmeaton@hotmail.com             paid 2018-02-08 00:42:57   
7  191569914709   litepiggy@yahoo.com.sg             paid 2018-02-07 14:58:31   
8  191569914708      alyssacmy@gmail.com             paid 2018-02-07 13:45:51   
9  191569914708      alyssacmy@gmail.com              NaN                 NaT   

  Fulfillment Status        Fulfilled at Accepts Marketing Currency  Subtotal  \
0          fulfilled 2018-02-08 03:59:01               yes      SGD     290.0   
1                NaN                 NaT               NaN      NaN       NaN   
2                NaN                 NaT               NaN      NaN       NaN   
3                NaN                 NaT               NaN      NaN       NaN   
4                NaN                 NaT               NaN      NaN       NaN   
5        unfulfilled                 NaT                no      SGD      38.0   
6        unfulfilled                 NaT               yes      SGD      48.0   
7        unfulfilled                 NaT               yes      SGD      75.0   
8        unfulfilled                 NaT               yes      SGD      42.0   
9                NaN                 NaT               NaN      NaN       NaN   

   Shipping  Taxes  Total     Discount Code  Discount Amount  \
0       0.0    0.0  290.0  5OFFe53b8c2fb9df              5.0   
1       NaN    NaN    NaN               NaN              NaN   
2       NaN    NaN    NaN               NaN              NaN   
3       NaN    NaN    NaN               NaN              NaN   
4       NaN    NaN    NaN               NaN              NaN   
5       5.5    0.0   43.5  5OFF5359ca0ded5b              5.0   
6       9.0    0.0   57.0               NaN              0.0   
7       0.0    0.0   75.0               NaN              0.0   
8       5.5    0.0   47.5  5OFF22c18122b276              5.0   
9       NaN    NaN    NaN               NaN              NaN   

                              Shipping Method          Created at  \
0                                         NaN 2018-02-08 03:59:00   
1                                         NaN 2018-02-08 03:59:00   
2                                         NaN 2018-02-08 03:59:00   
3                                         NaN 2018-02-08 03:59:00   
4                                         NaN 2018-02-08 03:59:00   
5                      2 Working-Day Delivery 2018-02-08 03:39:25   
6  Free Shipping above S$99 (ex.Taxes & Fees) 2018-02-08 00:42:56   
7              Free Delivery (2 Working Days) 2018-02-07 14:58:31   
8                      2 Working-Day Delivery 2018-02-07 13:45:51   
9                                         NaN 2018-02-07 13:45:51   

   Lineitem quantity                                      Lineitem name  \
0                  1    Resist Skin Restoring Moisturizer SPF 50 - 60ml   
1                  1  Resist Optimal Results Hydrating Cleanser - 19...   
2                  1  Resist Advanced Smoothing Treatment 10% AHA (G...   
3                  1                   Resist C15 Super Booster - 20 ml   
4                  1     Clinical Ceramide-Enriched Firming Moisturizer   
5                  1  Clear Acne Extra Strength Exfoliating Treatmen...   
6                  1  Calm Sensitive Daytime Moisturizer SPF 30 (Nor...   
7                 30  Resist Ultra-Light Super Antioxidant Concentra...   
8                  1  Skin Perfecting 2% BHA (Salicylic Acid) Liquid...   
9           

In [6]:


'''
comparison_table = sales_data_clean[['Email', 'Created at', 'discover_first_date', 'used_discover_already']]
comparison_table['discover_sales_lead_time'] = comparison_table['Created at'] - comparison_table['discover_first_date']
'''

# Filter for transactions that were made within a day of attempting Discover
same_day_purchase = sales_data_clean[(sales_data_clean['discover_sales_lead_time'] <= "1 days") & (sales_data_clean['discover_sales_lead_time'] > "0 days")]
same_day_purchases_count = len(same_day_purchase['Email'].unique())

# Perform a count and percentage of customer analysis 
# post_discover_all_sales_customer_count = len(post_discover_sales['Email'].unique())
percentage_same_day_purchase = same_day_purchases_count/post_discover_all_sales_customer_count
print("same_day_purchases_count:", same_day_purchases_count)
print("post_discover_all_sales_customer_count:", post_discover_all_sales_customer_count)
print("percentage_same_day_purchase: {:.2f}%".format(100*percentage_same_day_purchase))

# same_day_purchase.to_excel("same_day_purchase.xlsx")



NameError: name 'post_discover_all_sales_customer_count' is not defined

In [8]:
# Create two dataframes that include transactions before and on/after Discover launch (2017-09-09)
pre_discover_sales = sales_data_clean[sales_data_clean['Created at']< "2017-09-09"]
post_discover_sales = sales_data_clean[sales_data_clean['Created at']>= "2017-09-09"]

# Create a new dataframe from post_discover_sales
# that include only new users who took discover

only_users_from_post_discover_sales = post_discover_sales[post_discover_sales['Email'].isin(discover_first['email'].unique())]
only_users_from_post_discover_sales

# sales_data_clean[sales_data_clean['Email'] == 'jovi.kau@gmail.com']

# Create a new dataframe containing UNIQUE emails from post_discover_sales
post_launch_emails = pd.DataFrame(post_discover_sales['Email'].unique(), 
                                  columns=["Post Launch Emails"])

# Test if these emails are also found in pre_discover_sales email list
post_launch_emails['Exist before launch?'] = post_launch_emails.isin(pre_discover_sales['Email'].unique())

# Merge (aka Join) Discover Data with emails in post_launch_emails
# and create columns to indicate whether and when did the email tried discover
post_launch_emails = post_launch_emails.merge(right=discover_first[['email', 'timestamp_proper']], 
                                              left_on='Post Launch Emails', 
                                              right_on='email')

# Delete the extra email column created after joining
post_launch_emails = post_launch_emails.drop('email', axis=1)

# Rename the column into something more elucidatory
post_launch_emails = post_launch_emails.rename(index=str, columns={"timestamp_proper": "First Tried Discover"})

# get first transaction date of each of the "new" customers
# Remember that sales_data is reverse-chronologically indexed
post_launch_first_purchase = post_discover_sales[['Email', 'Created at']].drop_duplicates(subset='Email', keep='last')
print("Number of rows:", len(post_launch_first_purchase))
post_launch_first_purchase

post_launch_emails = post_launch_emails.merge(post_launch_first_purchase, 
                                              left_on="Post Launch Emails", 
                                              right_on="Email")
post_launch_emails = post_launch_emails.drop('Email', axis=1).rename(index=str, 
                                                                     columns={"Created at": "First Transaction Post Launch"})

post_launch_emails['Time To Buy'] = post_launch_emails['First Transaction Post Launch'] - post_launch_emails['First Tried Discover']

# Filter for purchases made in 1 day (if necessary)
# post_launch_emails = post_launch_emails[(post_launch_emails['Time To Buy']> "0 days") & (post_launch_emails['Exist before launch?'] == False)]

# Lead time to purchase for new customers after taking Discover
print(post_launch_emails[(post_launch_emails['Time To Buy'] > '0 days') & 
                         (post_launch_emails['Exist before launch?'] == False)]['Time To Buy'].describe())
print('Top 85% of time to buy:', post_launch_emails[(post_launch_emails['Time To Buy'] > '0 days') & 
                         (post_launch_emails['Exist before launch?'] == False)]['Time To Buy'].quantile(0.85))

post_launch_emails.head(5)

Number of rows: 2847
count                        346
mean     10 days 03:05:21.540462
std      21 days 16:32:44.176929
min              0 days 00:01:22
25%       0 days 00:19:19.500000
50%       0 days 09:24:47.500000
75%       9 days 00:29:25.250000
max            139 days 05:01:07
Name: Time To Buy, dtype: object
Top 85% of time to buy: 22 days 01:29:42.500000


Post Launch Emails  Exist before launch? First Tried Discover  \
0      jglyj82@gmail.com                 False  2018-02-07 12:54:54   
1    zarr.gyii@gmail.com                  True  2017-09-26 03:45:14   
2    ycobonpue@gmail.com                  True  2017-10-31 06:06:57   
3  karenkhor27@gmail.com                  True  2017-09-13 02:03:15   
4   boazruth76@yahoo.com                 False  2018-01-24 09:57:32   

  First Transaction Post Launch        Time To Buy  
0           2018-02-07 13:02:33    0 days 00:07:39  
1           2017-09-26 05:10:37    0 days 01:25:23  
2           2017-09-19 06:46:41 -42 days +00:39:44  
3           2017-09-18 07:12:43    5 days 05:09:28  
4           2018-01-11 04:18:16 -14 days +18:20:44

In [10]:
post_discover_sales

Name                         Email Financial Status  \
0        191569914712        gilly.glanville@me.com             paid   
1        191569914712        gilly.glanville@me.com              NaN   
2        191569914712        gilly.glanville@me.com              NaN   
3        191569914712        gilly.glanville@me.com              NaN   
4        191569914712        gilly.glanville@me.com              NaN   
5        191569914711       stephaniedata@yahoo.com             paid   
6        191569914710          hsmeaton@hotmail.com             paid   
7        191569914709        litepiggy@yahoo.com.sg             paid   
8        191569914708           alyssacmy@gmail.com             paid   
9        191569914708           alyssacmy@gmail.com              NaN   
10       191569914707             jglyj82@gmail.com             paid   
11       191569914707             jglyj82@gmail.com              NaN   
12       191569914706        yeoslfiona@hotmail.com             paid   
13       191569914706        yeoslfiona@hotmail.com              NaN   
14       191569914706        yeoslfiona@hotmail.com              NaN   
15       191569914706        yeoslfiona@hotmail.com              NaN   
16       191569914705       vinothini.112@gmail.com             paid   
17       191569914704        wailing_93@hotmail.com             paid   
18       191569914704        wailing_93@hotmail.com              NaN   
19       191569914703       jesse.dytioco@gmail.com             paid   
20       191569914703       jesse.dytioco@gmail.com              NaN   
21       191569914703       jesse.dytioco@gmail.com              NaN   
22     191569914-1758           zarr.gyii@gmail.com             paid   
23     191569914-1758           zarr.gyii@gmail.com              NaN   
24       191569914701     veron_chinron@hotmail.com             paid   
25       191569914701     veron_chinron@hotmail.com              NaN   
29     191569914-1755           ycobonpue@gmail.com             paid   
30       191569914697          zachery.ng@gmail.com             paid   
31       191569914696            weizen.t@gmail.com             paid   
32       191569914695  lady_portia3012@yahoo.com.sg             paid   
...               ...                           ...              ...   
10250     19156809825           Ft_hope@hotmail.com              NaN   
10251     19156809824             sonixrw@gmail.com             paid   
10252     19156809823           jivebunny@gmail.com             paid   
10253     19156809823           jivebunny@gmail.com              NaN   
10254     19156809823           jivebunny@gmail.com              NaN   
10255     19156809823           jivebunny@gmail.com              NaN   
10256     19156809823           jivebunny@gmail.com              NaN   
10257     19156809823           jivebunny@gmail.com              NaN   
10258     19156809822   chloe.yuting.wang@gmail.com             paid   
10259     19156809822   chloe.yuting.wang@gmail.com              NaN   
10260     19156809821         arya.nerker@gmail.com             paid   
10261     19156809821         arya.nerker@gmail.com              NaN   
10262     19156809821         arya.nerker@gmail.com              NaN   
10264     19156809819    jacjacqueline210@gmail.com             paid   
10265     19156809819    jacjacqueline210@gmail.com              NaN   
10266     19156809818           susannyyo@gmail.com             paid   
10267     19156809818           susannyyo@gmail.com              NaN   
10268     19156809817           susannyyo@gmail.com             paid   
10269     19156809817           susannyyo@gmail.com              NaN   
10270     19156809817           susannyyo@gmail.com              NaN   
10271     19156809817           susannyyo@gmail.com              NaN   
10272     19156809817           susannyyo@gmail.com              NaN   
10273     19156809817           susannyyo@gmail.com              NaN   
10274     19156809817           susannyyo@gmail.com      

In [9]:
post_launch_emails[(post_launch_emails['Time To Buy'] > '0 days') & 
                   (post_launch_emails['Time To Buy'] < '3 days') & 
                         (post_launch_emails['Exist before launch?'] == False)]['Time To Buy'].apply(lambda x: (x/pd.Timedelta(hours=1))).set_index('Post Launch Emails').cumsum()

plt.title('Count of hours required to purchase (Truncated to 72 hours)')
plt.xlabel('Number of hours')
plt.ylabel('Count');

AttributeError: 'Series' object has no attribute 'set_index'

In [11]:
# Average Order Value of _New_ Customers who made their first purchase after taking Discover

new_discover_sale_transaction = post_discover_sales[post_discover_sales['Email'].isin(new_customers_test['Post Launch Emails'].unique())]
# new_discover_sale_transaction.groupby(by=['Email', 'Created at'])['Total'].sum()
new_discover_sale_transaction['Total'].mean()


NameError: name 'new_customers_test' is not defined

### 50% of our new customers made a purchase within 10 hours after trying out Discover!
From the calculation above, we know that: 
1. there are 2847 first-time purchases since Discover was launched.
2. on average, it takes about **10 days** for a customer to make a purchase after using Discover
3. 50% of transactions since Discover was launched occured within **5 hours** of trying out Discover

In [12]:
# Obtain a list of transactions for NEW customers who made their first purchase WITHOUT trying Discover

count_with_discover = len(post_launch_emails[(post_launch_emails['Time To Buy'] >= '0 days') & 
                   (post_launch_emails['Exist before launch?'] == False)]['Post Launch Emails'].unique())
count_without_discover = len(post_launch_emails[(post_launch_emails['Time To Buy'] < '0 days') & 
                   (post_launch_emails['Exist before launch?'] == False)]['Post Launch Emails'].unique())

print('Unique count of new customers who purchased after trying Discover: {}'.format(count_with_discover))
print('Unique count of new customers who purchased without trying Discover: {}'.format(count_without_discover))
print('{:.1f}% of new customers made a purchase after trying out Discover'.format(count_with_discover/(count_with_discover + count_without_discover) * 100))

Unique count of new customers who purchased after trying Discover: 346
Unique count of new customers who purchased without trying Discover: 52
86.9% of new customers made a purchase after trying out Discover


In [13]:
post_launch_emails

Post Launch Emails  Exist before launch? First Tried Discover  \
0               jglyj82@gmail.com                 False  2018-02-07 12:54:54   
1             zarr.gyii@gmail.com                  True  2017-09-26 03:45:14   
2             ycobonpue@gmail.com                  True  2017-10-31 06:06:57   
3           karenkhor27@gmail.com                  True  2017-09-13 02:03:15   
4            boazruth76@yahoo.com                 False  2018-01-24 09:57:32   
5            bbggf.0901@gmail.com                  True  2017-10-27 17:38:14   
6         jeronblahblah@gmail.com                  True  2017-09-17 05:58:14   
7             ruienseah@gmail.com                 False  2018-02-06 08:16:14   
8         qingshuang111@gmail.com                 False  2017-10-01 15:37:11   
9           roseleenlua@gmail.com                  True  2018-01-16 05:28:25   
10       anisa.adhyanti@gmail.com                  True  2017-09-10 10:00:33   
11     parsheechandnani@gmail.com                 False  2017-10-31 06:32:51   
12              lukwern@gmail.com                 False  2017-10-14 09:05:21   
13   lohhuijuanrachel@hotmail.com                 False  2018-01-27 19:18:42   
14     shireengill.1991@gmail.com                  True  2018-02-05 14:31:05   
15            syshuka@hotmail.com                 False  2018-01-29 22:18:16   
16       imobviouslynot@gmail.com                 False  2018-02-04 19:43:12   
17            lyejiahui@gmail.com                  True  2017-09-09 16:07:52   
18            mabolight@gmail.com                 False  2017-10-21 04:13:03   
19          mirandalee88@live.com                  True  2017-11-01 02:48:44   
20             yyanti22@yahoo.com                  True  2017-10-07 14:34:53   
21      hazelwongshihui@gmail.com                  True  2017-09-10 06:38:23   
22    somu_luvhrithik@hotmail.com                  True  2017-10-06 05:00:39   
23              manasa@luxola.com                  True  2017-10-07 07:47:14   
24               pkly58@gmail.com                  True  2017-10-31 07:44:52   
25               tnxw@hotmail.com                  True  2017-10-31 08:24:13   
26        jessiejiaqiye@gmail.com                  True  2017-10-23 10:32:45   
27      tranlethuthuy94@gmail.com                 False  2017-10-09 05:03:47   
28          yenwei_chea@yahoo.com                  True  2018-02-03 14:16:00   
29   amanda_wong2003@yahoo.com.sg                 False  2018-02-03 01:17:14   
..                            ...                   ...                  ...   
698       giovanna.dioni@yahoo.it                  True  2017-09-15 08:54:43   
699         linmeinmein@gmail.com                 False  2017-09-16 05:52:45   
700          laipeiwong@gmail.com                 False  2017-09-10 04:23:57   
701         tram.hien27@gmail.com                 False  2017-09-16 05:04:43   
702           gyang13@hotmail.com                  True  2017-09-10 07:38:37   
703           epsh.1015@gmail.com                  True  2017-09-11 17:09:53   
704         plastickers@gmail.com                  True  2017-09-15 16:15:31   
705        syahirah17@hotmail.com                 False  2017-09-14 16:10:03   
706    social_fantasy@hotmail.com                 False  2017-09-15 02:46:01   
707      twofinerascals@gmail.com                 False  2017-09-14 17:54:55   
708           dionislim@gmail.com                 False  2017-09-14 14:31:55   
709         abc-kevin@live.com.sg                  True  2017-09-28 23:06:16   
710          kly0606@yahoo.com.sg                 False  2017-09-13 18:31:04   
711               mei_yin@live.cn                 False  2017-09-11 01:58:55   
712      joansoonjialin@gmail.com                  True  2017-09-13 18:33:47   
713         lyelye241@hotmail.com                 False  2017-09-13 03:03:12   
714             aprxlim@gmail.com                  True  2017-09-13 02:57:46   
715         xinyan_11@hotmail.com                 False  2017-10-31 05:52:54   
716          Ba

In [ ]:
count_with_discover = len(post_launch_emails[(post_launch_emails['Time To Buy'] >= '0 days') & 
                   (post_launch_emails['Exist before launch?'] == True)]['Post Launch Emails'].unique())
count_without_discover = len(post_launch_emails[(post_launch_emails['Time To Buy'] < '0 days') & 
                   (post_launch_emails['Exist before launch?'] == True)]['Post Launch Emails'].unique())

print('Distribution of time to buy for existing customers: \n{}'.format(post_launch_emails[(post_launch_emails['Time To Buy'] >= '0 days') & 
                         (post_launch_emails['Exist before launch?'] == True)]['Time To Buy'].describe()))
print('Top 85% of time to buy:', post_launch_emails[(post_launch_emails['Time To Buy'] > '0 days') & 
                         (post_launch_emails['Exist before launch?'] == True)]['Time To Buy'].quantile(0.85))


print('Unique count of existing customers who purchased after trying Discover: {}'.format(count_with_discover))
print('Unique count of existing customers who purchased without trying Discover: {}'.format(count_without_discover))
print('{:.1f}% of existing customers who made a purchase after trying out Discover'.format(count_with_discover/(count_with_discover + count_without_discover) * 100))

#### Obtain a list of unique emails of NEW customers who made their first purchase *without* taking Discover

In [ ]:

# Obtain a list of new customers who had made their first purchase post-launch without using Discover
new_cust_without_discover = post_launch_emails[(post_launch_emails['Time To Buy'] < pd.Timedelta('00:00:00')) & 
                                               (post_launch_emails['Exist before launch?'] == False)] 
                                                #['Post Launch Emails'].unique()

post_discover_sales[~post_discover_sales['Email'].isin(new_cust_without_discover['Post Launch Emails'].unique())]

# post_launch_emails was already defined earlier, and have not been altered in any way
first_sales_without_discover = post_discover_sales[~post_launch_emails.isin(pre_discover_sales['Email'].unique())]
first_sales_without_discover = first_sales_without_discover.dropna()
first_sales_without_discover

In [ ]:
type(new_cust_without_discover['Time To Buy'][0])

In [ ]:
new_cust_without_discover = post_launch_emails[post_launch_emails['Time To Buy'] < pd.Timedelta('00:00:00')]
new_cust_without_discover

## Average Order Value of _New_ Customers who made their first purchase _without_ taking Discover

In [ ]:
new_non_discover_sale_transaction = post_discover_sales[~post_discover_sales['Email'].isin(new_customers_test['Post Launch Emails'].unique())]
new_non_discover_sale_transaction['Total'].mean()

# Comparison

In [ ]:
table_to_compare = pivot_discover_first
source_table_to_compare = sales_data_clean

In [ ]:
multiindex_output = table_to_compare.set_index(keys=['email', 'SKU'])
multiindex_output

In [ ]:
check_presence = multiindex_output.isin(source_table_to_compare[['Email', 'Lineitem sku']])
check_presence[check_presence['step_no'] == True]

### Following code aims to examine the proportion of SKUs *actually* recommended to the customers
#### This helps to examine the trend of skin type (better to get from raw data for this)
"unpivot" ("melt" in Pandas term) in discover_first[result_SKUs].
### The objective is to allow for easier indexing to search for SKUs in sales_data_clean

In [ ]:
# pivot_discover_first = discover_first.unstack(index='email', columns = ['result_0', 'result_1',
#       'result_2', 'result_3', 'result_4', 'result_5', 'result_6', 'result_7',
#       'result_8', 'result_9'])
pivot_discover_first = pd.melt(discover_first, 
                               id_vars='email',
                               value_vars=list(discover_first.columns[-11:-1]),
                               var_name='step_no',
                               value_name= 'SKU')

In [ ]:
pivot_discover_first.sort_values(['email', 'step_no'], inplace=True)
# pivot_discover_first.to_excel("pivot_discover_first.xlsx")
# pivot_discover_first
discover_first.columns[-10:-1]
discover_first[discover_first['email'] == "yuleandra21@gmail.com"]
# same_day_purchase.groupby(by=['Email','Lineitem sku'])['Lineitem sku'].count()
#pivot_discover_first['Purchased?'] = pivot_discover_first['email'].map(same_day_purchase.set_index(['Email', 'Lineitem sku'])[])

'''Faced issues creating a frequency plot for SKUs under each email based on sales data
Abandoned using jupyter, tried on excel instead, also gave up subsequently
(Might need to use for loop with two filter criteria, but could not figure out how)'''